DM-30977: Add python log handler to store LogRecord as they come in

timj opened the pull request:
These records can then be serialized in butler datastore.  This is assuming that lsst/log#55 is merged which will also need changes in daf_butler for initializing logging such that we no longer forward everything to lsst.log. Will need changes to ctrl_mpexec as well.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add python log handler class to store LogRecord as they come in

These records can then be serialized in butler datastore.

andy-slac commented:
DM-30996 will most likely add `MDC` attribute to `LogRecord` which will be a dict-like object (or maybe SimpleNamespace but it should support dict methods too). I think we want to save its contents too in some portable form.

timj commented:
The new `ButlerLogRecord` can be as flexible as we want it to be. I can definitely make it scan the `logging.LogRecord` for additional dict-like objects and add them in to `ButlerLogRecord.extra` or add them into the expanded message.

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -21,11 +21,185 @@
+class ButlerLogRecord(BaseModel):
+    """A model representing a `logging.LogRecord`.
+
+    A `~logging.LogRecord` always uses the current time in its record
+    when recreated and that makes it impossible to use it as a
+    serialization format. Instead have a local representation of a
```
You can overwrite LogRecord attributes after creating it, you just need to know which ones, and there is a `logging.makeLogRecord()` method which looks like it was specifically designed for supporting LogRecord serialization (probably as a dict). Pickling LogRecod directly is probably not the right thing to do, in particular new MDC attribute needs some care, you probably want to serialize it a as a simple dict, to avoid dependency on `log.lsst.MDCDict` class.

timj replied:
Right, that's all why I created `ButlerLogRecord` so that we can have the flexibility to deal with MDC.

timj committed:
Add MDC to the ButlerLogRecord definition

timj committed:
Add formatter and storage class definition for ButlerLogRecords

timj committed:
Add exc_info to the ButlerLogRecord

There is a question as to whether this should be attached to
the main message, kept as a separate string, or stored
separately as a list of strings.

For now it is stored as a separate single string.

timj committed:
Add JsonFormatter for formatting log messages to JSON

This can be used to directly stream LogRecords in JSON format.

timj committed:
Allow ButlerLogRecords to be sliced

Returning a new ButlerLogRecords instance.

timj committed:
Support new log record stream format

One record per line rather than a single JSON model per file.

timj committed:
Improve error message when ingest not implemented by datastore

timj committed:
Tell pydantic to skip unset/default values in json for ButlerLogRecord

timj committed:
Add --log-file option to command line tooling

timj committed:
Move json log file reader to ButlerLogRecords

This makes it available to test code and to other users without
having to limit reading to butler formatter.

timj committed:
Update the comment relating to log test

andy-slac commented on python/lsst/daf/butler/cli/cliLog.py:
```diff
@@ -152,8 +161,22 @@ def record_factory(*args, **kwargs):
 
+        # Set up the file logger
+        if log_file:
+            handler = logging.FileHandler(log_file, delay=True)
```
Is `delay` useful? May be better to fail early if file cannot be open? Also there could be non-trivial interaction with potential `chdir` use.

timj replied:
Maybe not. I ended up adding this to see if it made a difference when debugging a problem with not being able to close files, but I'll remove it now.

andy-slac commented on python/lsst/daf/butler/cli/opt/options.py:
```diff
@@ -113,6 +113,12 @@ def makeCollectionTypes(context, param, value):
+log_file_option = MWOptionDecorator("--log-file",
+                                    default=None,
+                                    type=MWPath(file_okay=True, dir_okay=False, writable=True),
+                                    help="File to write log messages. If .json JSON log records will be"
```
`If .json` -> maybe `If ends with .json then ...`?

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+            record_dict["MDC"] = dict(record.MDC)  # type: ignore
+        else:
+            # Use an empty MDC to prevent format failures.
+            record_dict["MDC"] = defaultdict(str)
```
Maybe

    record_dict["MDC"] = getattr(record, "MDC", defaultdict(str))

timj replied:
Actually that's not quite the same because the original code makes a point of copying `record.MDC` into a new dict so that later changes to the MDC won't change the record after the fact. With copying I think it needs to be:

```
record_dict["MDC"] = dict(getattr(record, "MDC", {})
```
This does make the formatting a touch problematic if `LABEL` is not in MDC but I'm not sure what to do about that anyway since as written the record constructed from json won't be using defaultdict either.

Also, since this is a pydantic model it's a bit tricky to add keys at the top level so doing `record_dict(**record.MDC)` isn't really going to work.

andy-slac replied:
So if we need more dynamic model for this record type we cannot use pydantic? Like with any other persistent data we need to be careful with schema changes, at some point we can decide to add/remove attributes to this class and then we need to think about compatibility between new code and what is already stored in butler (or old code and data saved by new release). Maybe we need this class to be a simple dict with arbitrary set of keys (though it would just move compatibility issue to a different level)?

timj replied:
Depends what you mean by dynamic. At the moment with pydantic I've set up the model so that all the LogRecord items are included (I'm keeping the exc_info text distinct from the message itself for now) and all the arbitrary stuff goes into the `MDC` attribute which can accept anything. That seems flexible enough to me. Maybe I should rename MDC to `extra`?  I've never ingested records into LogStash so I don't know how well it will deal with the `MDC` dict or whether it really wants `record.LABEL` to be used and the `MDC` hierarchy missing completely.  The `LogRecord` formatting of `MDC[LABEL]` is not the same code as a the `ButlerLogRecord` serialization so I can convert the `MDC` part of the dict to a defaultdict at that point when calling `ButlerLogRecord.format()`

andy-slac replied:
By dynamic I mean that someone could decide later that using `extra` keyword to logging methods is a great idea (in addition to MDC) and wants to save all those LogRecord attributes in the same JSON format as top-level keys. That set of attributes is potentially unbounded. Packing all of them into `MDC` or one `extra` attribute in this class may or may not work depending on how they are used, I guess we can only discover that later. We could probably reserve `extra` attribute for that future extension and hope that nesting all those extra attributes into one dictionary will work for everyone.

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+
+        # A stream of records with one record per line.
+        if first_char != "{":
+            raise RuntimeError(f"Unrecognized JSON log format. First lines is '{first_line}'")
```
typo: lines is

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+        first_char = str(first_line[0])
+        if first_char == "[":
+            # This is a ButlerLogRecords model serialization.
+            all = first_line + stream.read()
+            return cls.parse_raw(all)
```
Does it need whole file content in memory to parse? Could be an issue with large files. But I guess to write it in this format you also need everything in memory.

timj replied:
Yes -- this is all working on the assumption that at some point you had it all in a `ButlerLogRecords` pydantic model anyhow. I've been able to read half a million records and it's a bit slow but it works. I have been wondering about adding simple filtering on read (at least for the streaming read) to filter by logger name and log level).

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+    def __reversed__(self) -> Iterator[ButlerLogRecord]:
+        return self.__root__.__reversed__()
+
+    def __delitem__(self, index: int) -> None:
```
Can `index` be a slice here too?

timj replied:
Maybe. I hadn't thought about it. Will look.

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+        elif isinstance(record, LogRecord):
+            record = ButlerLogRecord.from_record(record)
+        else:
+            raise ValidationError(f"Can only not append item of type {type(record)}")
```
`Can only not append`? 🙂

andy-slac commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -94,3 +105,270 @@ def MDCRemove(cls, key: str) -> None:
+        self.records.append(record)
+
+
+class JsonFormatter(Formatter):
```
There is another class with the same name in this package, maybe call this `JsonLogFormatter`?

timj replied:
Oh yes. Luckily we never export datastore formatters. I have been wondering whether I should export any of these classes -- forcing pipetask to pull them in explicitly from `lsst.daf.butler.core.loggng`.

andy-slac commented on python/lsst/daf/butler/formatters/logs.py:
```diff
@@ -0,0 +1,121 @@
+
+    This is a naive implementation that treats everything as a pydantic.
+    model.  In the future this may be changed to be able to read
+    ``ButlerLogRecord`` one at time from the file and return a subset
```
Can do single backticks?

andy-slac commented on python/lsst/daf/butler/formatters/logs.py:
```diff
@@ -0,0 +1,121 @@
+        Notes
+        -----
+        Can read two forms of JSON log file. It can read a full JSON
+        document created from `ButlerLogRecords`, or a stream off standalone
```
typo: off

andy-slac commented on python/lsst/daf/butler/formatters/logs.py:
```diff
@@ -0,0 +1,121 @@
+        path : `str`
+            Path to use to open JSON format file.
+        pytype : `class`, optional
+            Not used by this implementation.
```
Is this true?

timj replied:
No. I could not use it but since it's there I am using it. Thanks.

andy-slac commented on python/lsst/daf/butler/formatters/logs.py:
```diff
@@ -0,0 +1,121 @@
+        serializedDataset : `bytes`
+            Bytes object to unserialize.
+        pytype : `class`, optional
+            Not used by this implementation.
```
Same here

andy-slac commented on python/lsst/daf/butler/formatters/logs.py:
```diff
@@ -0,0 +1,121 @@
+            return pytype.parse_raw(serializedDataset)
+
+        if not serializedDataset.startswith(b"{"):
+            raise RuntimeError(f"These bytes do not look like JSON -- starts with '{serializedDataset[0]}'")
```
I think this will fail with different exception if `serializedDataset` is empty

timj replied:
Good point. I've been wondering whether to also move this to `ButlerLogRecords` as `.from_raw()` since at the moment the formatter has the byte reading version but the class itself has the file reading version.

andy-slac commented on python/lsst/daf/butler/formatters/logs.py:
```diff
@@ -0,0 +1,121 @@
+            raise RuntimeError(f"These bytes do not look like JSON -- starts with '{serializedDataset[0]}'")
+
+        # Filter out blank lines.
+        records = [ButlerLogRecord.parse_raw(line) for line in serializedDataset.split(b"\n") if line]
```
`split()` is probably not super-efficient space-wise for large data volumes.

timj replied:
Is there a version of `.split()` somewhere that returns an iterator rather than a list? Or should I walk through the string myself looking for `\n` ?  I think this being a list means that we have all the bytes in memory, and all of them again in a list and all of them accumulating into the output list... On the plus size Datastore has an internal limit as to how big a dataset is allowed to be before using fromBytes.

timj replied:
The internet is full of people asking for `str.isplit()` and it never happening.

andy-slac replied:
I don't think there is an iterator version of split() in the standard library, but writing a simple generator method for it would be trivial.

andy-slac reviewed (approved):
Looks good, few minor comments.

timj committed:
Note that ButlerLogRecords.__delitem__ supports slicing and test it

timj committed:
Move byte string parser to ButlerLogRecords

Also unify the format detection code rather than repeating it
in stream and byte handling.

timj committed:
Implement a str.split generator to reduce memory overhead

When splitting a potentially enormous string we want to avoid
creating a list with potentially millions of elements.

timj committed:
Significantly expand test coverage of ButlerLogRecord

timj commented on python/lsst/daf/butler/core/logging.py:
```diff
@@ -21,16 +21,63 @@
 logging.addLevelName(VERBOSE, "VERBOSE")
 
 
+def _isplit(string: Union[bytes, str], sep: Optional[Union[bytes, str]]) -> Iterator[Union[bytes, str]]:
```
@andy-slac This is what I came up with for a split generator. I initially wrote:

```diff
diff --git a/python/lsst/daf/butler/core/logging.py b/python/lsst/daf/butler/core/logging.py
index 2a6ac1ab..ede1e6fe 100644
--- a/python/lsst/daf/butler/core/logging.py
+++ b/python/lsst/daf/butler/core/logging.py
@@ -41,6 +41,30 @@ _LONG_LOG_FORMAT = "{levelname} {asctime} {name} {filename}:{lineno} - {message}
 logging.addLevelName(VERBOSE, "VERBOSE")
 
 
+def _isplit(string: Union[bytes, str], sep: Optional[Union[bytes, str]]) -> Iterator[Union[bytes, str]]:
+    """Generator version of split to avoid making potentially large list.
+
+    Notes
+    -----
+    Python `str` do not support `memoryview` so slicing the string on each
+    iteration to chop the front off will copy the entire remainder of the
+    string. This implementation therefore keeps track of the start and end
+    positions.
+    """
+    if sep is None:
+        sep = " " if isinstance(string, str) else b" "
+    end = None
+    while end != -1:
+        # the type signature above does not explain that the string and
+        # sep types must match up.
+        end = string.find(sep)  # type: ignore
+        if end != -1:
+            yield string[:end]
+            string = string[end + 1:]
+    yield string
+    return
+
```
but then learned that `string = string[end:]` generates a copy so you would still end up with double the memory being required if the string is large.

mypy hates the current implementation and I'm not entirely sure what I have to do to tell mypy:

* All the parameters given and returned are the same type and not arbitrary combinations.
* That the result of `string[0]` is an `int` if `string` is `bytes` -- mypy really gets upset about the line that converts the int back to bytes again.

andy-slac replied:
What about something like 

    T = TypeVar('T', str, bytes)

    def _isplit(string: T, sep: T) -> Iterator[T]:
        begin = 0
        while True:
            end = string.find(sep, begin)
            if end == -1:
                yield string[begin:]
                return
            yield string[begin:end]
            begin = end + 1

timj replied:
Ha. Serves me right for not reading the `str.find()` docs completely and so not noticing the `begin` parameter. Thanks.

timj committed:
Move log factory addition to ButlerMDC

This allows other test code to use it and encapsulates the
access to the _MDC private dict.

timj committed:
Add tests for MDC in log messages

timj committed:
Allow --log-file to be used to specify multiple output files

timj committed:
Add --no-log-tty flag to allow disabling of terminal output

timj committed:
Improve MDC test coverage

timj committed:
Fix ButlerLogRecords.log_format and test it

timj committed:
Add test that blank lines in the stream log format are ignored

timj committed:
Restrict what is exported from logging by default

Only make ButlerLogRecords public by default since that is
the type stored in datastore.

Also rename JsonFormatter to JsonLogFormatter to avoid
confusion with the datastore formatter of the same name.

timj committed:
Add news fragment

timj committed:
Fix config replay of log options

timj committed:
Add ButlerMDC context manager

This temporarily changes the MDC for logging.
